src/esp/metadata/attributes/SceneAttributes.h

```diff
@@ -450,6 +461,60 @@ class SceneAttributes : public AbstractAttributes {
  public:
   ESP_SMART_POINTERS(SceneAttributes)
 };  // class SceneAttributes
+
+template <class T>
+void SceneAttributes::setObjectInstanceAttrInternal(
+    std::shared_ptr<T>& objInst,
+    std::deque<int>& availableIDs,
+    const std::shared_ptr<Configuration>& objInstConfig,
+    const std::string& objInstNamePrefix) {
+  // get subconfig for articulated object instances, add this ao instance as a
+  // set id
+  if (!availableIDs.empty()) {
+    // use saved value and then remove from storage
+    objInst->setID(availableIDs.front());
+    availableIDs.pop_front();
+  } else {
+    // use size of container to set ID
+    objInst->setID(objInstConfig->getNumSubconfigEntries());
+  }
+  // get last key
+  objInstConfig->setSubconfigPtr<T>(
+      Cr::Utility::formatString("{:.05d}_{}{}", objInst->getID(),
+                                objInstNamePrefix,
+                                objInst->getSimplifiedHandle()),
+      objInst);
+}
+
+template <class T>
+std::vector<std::shared_ptr<const T>>
+SceneAttributes::getObjectInstanceAttrInternal(
+    const std::shared_ptr<Configuration>& objInstConfig) const {
+  std::vector<std::shared_ptr<const T>> res{};
+  // pair of begin/end const iters through subconfig of given name
+  int numSubconfigs = objInstConfig->getNumSubconfigEntries();
+  if (numSubconfigs == 0) {
+    return {};
+  }
+  res.reserve(numSubconfigs);
+  auto objInstIter = objInstConfig->getSubconfigIterator();
+  // iterate through subconfig entries, casting appropriately and adding to
+  // map if cast successful
+  for (auto objIter = objInstIter.first; objIter != objInstIter.second;
+       ++objIter) {
+    auto obj = objIter->second;
+    if (std::shared_ptr<T> objPtr = std::dynamic_pointer_cast<T>(obj)) {
+      res.push_back(objPtr);
```

Comment:
```suggestion
      res.push_back(std::move(objPtr));
```
or better yet
```suggestion
      res.emplace_back(std::move(objPtr));
```

Reply:
I want a copy to save in the object that it creates.

Reply:
@jturner, it's a shared ptr though, no? So that should just more efficiently transfer the reference.

Reply:
Yeah, you're right. Thanks :)